### PR TITLE
Strengthen the pattern condition for identifying RC temp files

### PR DESCRIPTION
**Pull request by narasamdya**
DFA occurs on `d:\dbs\el\esi\obj\x86fre\termsrv\rdms\server\service\objfre\i386\rdms.lce`

The directory `d:\dbs\el\esi\obj\x86fre\termsrv\rdms\server\service\objfre\i386` is a shared opaque output directory. 

The write to that file by RC was ignored by sandboxed process reporting because of a weak identification pattern. Thus, subsequent reads/probes to that file by the same pip are considered access violation.

[AB#1584975](https://dev.azure.com/mseng/708e929f-6bd5-415a-8daf-25b1dac08dd8/_workitems/edit/1584975)

### CODE REVIEW

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs [hunk 1]
```diff
@@ -2619,6 +2624,41 @@ private SpecialProcessKind GetProcessKind(AbsolutePath processOverride)
...
+            char c3 = fileName[beginCharIndex + 3];
+            if (c3.ToUpperInvariantFast() != 'L')
+            {
+                return false;
+            }
+
+            return true;
+        }
+
         private static bool StringLooksLikeRCTempFile(string fileName)
```
Comment (mpysson) [inline Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: >StringLooksLikeRCTempFile [](start = 28, length = 25)

I'm really not a fan of these special cased exceptions in SandboxedProcessPipExecutor. You're going the right direction in making this one less likely to be exercised but I'd love to move more in the direction of not utilizing them at all. For shared opaques, it seems like we shouldn't really even need them

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs [hunk 2]
```diff
@@ -2619,6 +2624,41 @@ private SpecialProcessKind GetProcessKind(AbsolutePath processOverride)
             return s_specialTools.TryGetValue(toolName.ToLowerInvariant(), out SpecialProcessKind kind) ? kind : SpecialProcessKind.NotSpecial;
         }
 
+        private static bool StringLooksLikeBuildXLTestTempFile(string fileName)
```
Comment (mpysson) [inline Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: >StringLooksLikeBuildXLTestTempFile [](start = 28, length = 34)

why is this addition necessary?

**Comment by mpysson** [comment]
> bool StringLooksLikeRCTempFile(PCPathChar str, size_t str_length)

Can we pinvoke the native code instead of duplicating it in C#?

---
Refers to: Public/Src/Sandbox/Windows/DetoursServices/StringOperations.cpp:238 in f45b785. [](commit_id = f45b785446a7724a533cfef4acf873268063a251, deletion_comment = False)

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/SharedOpaqueDirectoryTests.cs [hunk 3]
```diff
@@ -1257,6 +1257,36 @@ public void AbsentPathProbeInUndeclaredOpaquesUnsafeModeCachedPip(bool forceDepe
...
+            AbsolutePath sharedOpaqueDirPath = AbsolutePath.Create(Context.PathTable, sharedOpaqueDir);
+            DirectoryArtifact sharedOpaqueDirArtifact = DirectoryArtifact.CreateWithZeroPartialSealId(sharedOpaqueDirPath);
+            FileArtifact outputInSharedOpaqueDir = CreateOutputFileArtifact(root: sharedOpaqueDir, prefix: "sod-file");
+            FileArtifact sourceFile = CreateSourceFile();
+
+            FileArtifact whitelistFile = CreateSourceFile();
+            global::BuildXL.Utilities.Configuration.Mutable.FileAccessWhitelistEntry entry = new global::BuildXL.Utilities.Configuration.Mutable.FileAccessWhitelistEntry()
+            {
+                Value = "testValue",
+                PathFragment = ArtifactToString(whitelistFile),
```
Comment (mpysson) [inline Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/SharedOpaqueDirectoryTests.cs]: >whitelistFile [](start = 48, length = 13)

It doesn't look like the test pip is interacting with this file. Maybe I'm missing the point of the test?

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs (hunk 1 above)
Comment (narasamdya) [reply Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: I know you're not a big fan of this exception. We can add a field in the configuration file describing the tool and the paths to ignore. But I guess that's beyond the purpose of this PR. We can add a user story to have a proper implementation.

Yes, shared opaque directories do not need this change in Detours. This change is for completeness.

---
In reply to: [314459791](https://github.com/microsoft/BuildXL/pull/741#discussion_r314459791) [](ancestors = 314459791)

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs (hunk 2 above)
Comment (narasamdya) [reply Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: There's no unit test for this special exception. I just want to add one to exercise the feature.

---
In reply to: [314459941](https://github.com/microsoft/BuildXL/pull/741#discussion_r314459941) [](ancestors = 314459941)

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/SharedOpaqueDirectoryTests.cs (hunk 3 above)
Comment (narasamdya) [reply Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/SharedOpaqueDirectoryTests.cs]: Good catch. At first I thought that the accesses were not reported because I didn't have whitelist entries. I'm removing it.

---
In reply to: [314461083](https://github.com/microsoft/BuildXL/pull/741#discussion_r314461083) [](ancestors = 314461083)

**Comment by narasamdya** [comment]
> bool StringLooksLikeRCTempFile(PCPathChar str, size_t str_length)

We can. But I'm not sure about the pinvoke on Mac side. Let's create a user story for that as such a change is beyond the purpose of this PR.

---
In reply to: [521764276](https://github.com/microsoft/BuildXL/pull/741#issuecomment-521764276) [](ancestors = 521764276)

---
Refers to: Public/Src/Sandbox/Windows/DetoursServices/StringOperations.cpp:238 in f45b785. [](commit_id = f45b785446a7724a533cfef4acf873268063a251, deletion_comment = False)

File: Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs (hunk 2 above)
Comment (mpysson) [reply Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs]: I'm not following. This adds a new exception rule for a new process into the product code. Why does adding a unit test necessitate that?

---
In reply to: [314472403](https://github.com/microsoft/BuildXL/pull/741#discussion_r314472403) [](ancestors = 314472403,314459941)

**Comment by mpysson** [comment]
> bool StringLooksLikeRCTempFile(PCPathChar str, size_t str_length)

None of these rules apply for mac builds since they are just special cases for windows tools

---
In reply to: [521778803](https://github.com/microsoft/BuildXL/pull/741#issuecomment-521778803) [](ancestors = 521778803,521764276)

---
Refers to: Public/Src/Sandbox/Windows/DetoursServices/StringOperations.cpp:238 in f45b785. [](commit_id = f45b785446a7724a533cfef4acf873268063a251, deletion_comment = False)
